core/src/cubos/core/thread_pool.cpp

```diff
@@ -0,0 +1,63 @@
+#include <cubos/core/thread_pool.hpp>
+
+using namespace cubos::core;
+
+ThreadPool::ThreadPool(size_t numThreads)
+{
+    this->threads.reserve(numThreads);
+    this->numTasks = 0;
+    this->stop = false;
+
+    for (size_t i = 0; i < numThreads; i++)
+    {
+        this->threads.emplace_back([this]() {
+            while (true)
+            {
+                std::function<void()> task;
+
+                {
+                    std::unique_lock<std::mutex> lock(this->mutex);
+
+                    // Wait until a task is available, or the thread pool is being destroyed.
+                    this->new_task.wait(lock, [this]() { return this->stop || !this->tasks.empty(); });
+                    if (this->stop && this->tasks.empty())
+                        return; // Thread pool is being destroyed, and there are no more tasks to execute.
+                    task = std::move(this->tasks.front());
+                    this->tasks.pop_front();
+                    this->numTasks += 1; // A task is being executed.
+                }
+
+                task();
+                this->numTasks -= 1; // Task has finished executing.
```

Comment:
This doesn't look like it's being protected with a lock.

Reply:
I doesn't need to, thats the async part.
If you're referring to the `numTasks` variable, its atomic so its okay to decrement it without the mutex.